build.sh

```diff
@@ -0,0 +1,112 @@
+# various artifacts for the Pbench product.  It is intended to be run from the
+# root directory of a Git branch checkout.
+
+# The following variables will normally be defined as credentials in the Jenkins
+# environment.  For developer use of this script, they should be defined in the
+# shell environment; otherwise, they will be given default values, here.
+#
+#  COPR_CONFIG - the path to the configuration file for the COPR account
+#  PODMAN_USR - the username for the container registry account
+#  PODMAN_PSW - the password for the container registry account
+#
+PODMAN_USR=${PODMAN_USR-"pbench+mr_jenkins"}
+PODMAN_PSW=${PODMAN_PSW-${DEFAULT_PODMAN_PASSWORD}}
+COPR_CONFIG=${COPR_CONFIG:-"${HOME}/.config/copr"}
+
+
+# Function to read an initialization/configuration file and print the value of
+# the specified key in the specified section.  Borrowed from
+# https://stackoverflow.com/questions/49399984/parsing-ini-file-in-bash/49400353
+function getIni() {
+  local f=${1}  # Initialization/configuration file name
+  local s=${2}  # Section name
+  local o=${3}  # Option name
+
+  if [[ ! -r ${f} ]]; then
+    echo "Missing or bad COPR config file:  ${f}" >&2
+    exit 1
+  fi
+
+  local rv=$(awk '/^\[.*\]$/{obj=$0}/=/{print obj $0}' ${f} | \
+    sed -En "/${s}]${o}/s/^.*=[[:space:]]*//p" | \
+    tail -n 1)
+
+  if [[ -z ${rv} ]]; then
+    echo "Missing '${o}' in COPR config file, '${f}'" >&2
+    exit 1
+  fi
+
+  echo ${rv}
+}
+
+
+COPR_USER=$(getIni ${COPR_CONFIG} copr-cli username)
+COPR_URL=$(getIni ${COPR_CONFIG} copr-cli copr_url)
+
+if [[ ${COPR_URL} =~ "copr.fedorainfracloud.org" ]]; then
+  URL_PREFIX=https://download.copr.fedorainfracloud.org/results/${COPR_USER}
+elif [[ ${COPR_URL} =~ "copr.devel.redhat.com" ]]; then
+  URL_PREFIX=http://coprbe.devel.redhat.com/results/${COPR_USER}
+else
+  echo "Unexpected COPR_URL: '${COPR_URL}'" >&2
+  exit 1
+fi
+
+# Install the linter requirements and add them to the PATH.
+export PATH=${HOME}/.local/bin:${PATH}
+python3 -m pip install --user -r lint-requirements.txt
+
+# If this script is run in a container and the user in the container doesn't
+# match the owner of the Git checkout, then Git objects; these config settings
+# prevent the complaints.
+GITTOP=$(git rev-parse --show-toplevel 2>&1 | head -n 1)
+if [[ ${GITTOP} =~ "fatal: unsafe repository ('/home/root/pbench'" ]] ; then
+	git config --global --add safe.directory /home/root/pbench
+	git config --global --add safe.directory /home/root/pbench/agent/stockpile
+	GITTOP=$(git rev-parse --show-toplevel)
+fi
+
+# Install the Dashboard dependencies, including the linter's dependencies and
+# the unit test dependencies.
+( cd dashboard && npm install )
+
+# Test for code style and lint
+black --check .
+flake8 .
+( cd dashboard && npx eslint "src/**" --max-warnings 0 )
+
+# Run unit tests
+tox                                     # Agent and Server unit tests and legacy tests
```

Comment:
I'm not sure what's going on here, but somehow _you're only running the "agent" tests_. I glanced through the Cobertura report and noticed 0% coverage for all the server code, and the console log confirms this.

Cobertura is also voluminously complaining about all the (pypi) module source files it can't find, which is annoying. I don't recall seeing these before when it couldn't find *any* of the source files, so I'm not sure what changed.

Reply:
I'll look into what's going on with the Server tests.  However, take a look at the log for [PR-2891](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/view/change-requests/job/PR-2891/17/console) (which I think would be familiar to you 😉) -- isn't it full of the same Cobertura complaints?

I agree that the diagnostics in the log are annoying; even worse are all of the non-project lines in the coverage report.  I assume that there is some way to get Cobertura to report only on _our_ code, but I haven't looked into that.  (Just getting the source working in the report was enough of an achievement for the nonce!)